crates/router/src/core/api_locking.rs

```diff
@@ -0,0 +1,155 @@
+use std::fmt::Debug;
+
+use actix_web::rt::time as actix_time;
+use error_stack::{IntoReport, ResultExt};
+use redis_interface as redis;
+use router_env::{instrument, logger, tracing};
+
+use super::errors::{self, RouterResult};
+use crate::routes::{app::AppStateInfo, lock_utils};
+
+pub const API_LOCK_PREFIX: &str = "API_LOCK";
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum LockStatus {
+    // status when the lock is acquired by the caller
+    Acquired, // [#2129] pick up request_id from AppState and populate here
+    // status when the lock is acquired by some other caller
+    Busy,
+}
+
+#[derive(Clone, Debug)]
+pub enum LockAction {
+    // Sleep until the lock is acquired
+    Hold { input: LockingInput },
+    // Queue it but return response as 2xx, could be used for webhooks
+    QueueWithOk,
+    // Ignore it and return response as 2xx, could be used for webhooks
+    IgnoreWithOk,
+    // Return Error
+    Drop,
+    // Locking Not applicable
+    NotApplicable,
+}
+
+#[derive(Clone, Debug)]
+pub struct LockingInput {
+    pub unique_locking_key: String,
+    pub api_identifier: lock_utils::ApiIdentifier,
+}
+
+impl LockingInput {
+    fn get_redis_locking_key(&self, merchant_id: String) -> String {
+        format!(
+            "{}_{}_{}_{}",
+            API_LOCK_PREFIX, merchant_id, self.api_identifier, self.unique_locking_key
+        )
+    }
+}
+
+impl LockAction {
+    #[instrument(skip_all)]
+    pub async fn perform_locking_action<A>(self, state: &A, merchant_id: String) -> RouterResult<()>
+    where
+        A: AppStateInfo,
+    {
+        match self {
+            Self::Hold { input } => {
+                let redis_conn = state
+                    .store()
+                    .get_redis_conn()
+                    .change_context(errors::ApiErrorResponse::InternalServerError)?;
+
+                let redis_locking_key = input.get_redis_locking_key(merchant_id);
+                let delay_between_retries_in_milliseconds = state
+                    .conf()
+                    .lock_settings
+                    .delay_between_retries_in_milliseconds;
+                let redis_lock_expiry_seconds =
+                    state.conf().lock_settings.redis_lock_expiry_seconds;
+                let lock_retries: u32 = ((redis_lock_expiry_seconds * 1000)
```

Comment:
Will move this logic to the place where Settings is initialised and use check math functions here instead of direct operators.